### PR TITLE
feat: add cli commands to add/remove sources for multi-source applications

### PR DESCRIPTION
**Pull request by ishitasequeira**
The PR focusses on 3 commands:
* `argocd app get` - The current app get command shows details only regarding the first source in the header as summary. The PR prints summary for all the sources in the header.
* `argocd app add-source` - Allows users to append a source to the list of sources for a multi-source app.
* `argocd app remove-source` - Allows users to remove a specific source from the list of sources using index of the source.

Screenshots of commands:
* Single Source  for `argocd app get`
image

* `argocd app add-source` and Multi-source `argocd app get`
image

* `argocd app remove-source` and Multi-source `argocd app get`
image

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: cmd/argocd/commands/app.go [hunk 1]
```diff
@@ -2780,3 +2797,127 @@ func NewApplicationPatchCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+
+				_, err = appIf.UpdateSpec(ctx, &application.ApplicationUpdateSpecRequest{
+					Name:         &app.Name,
+					Spec:         &app.Spec,
+					Validate:     &appOpts.Validate,
+					AppNamespace: &appNs,
+				})
+				errors.CheckError(err)
+
+				fmt.Printf("application '%s' updated successfully\n", app.ObjectMeta.Name)
```
Comment (jannfis) [inline cmd/argocd/commands/app.go:2849]:
```suggestion
				fmt.Printf("Application '%s' updated successfully\n", app.ObjectMeta.Name)
```

I feel that capitalization is important for UX.

File: cmd/argocd/commands/app.go [hunk 2]
```diff
@@ -2780,3 +2797,127 @@ func NewApplicationPatchCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+			ctx := c.Context()
+
+			if len(args) != 1 {
+				c.HelpFunc()(c, args)
+				os.Exit(1)
+			}
+
+			if source_index < 0 {
+				errors.CheckError(fmt.Errorf("index value of source cannot be less than 0"))
+				os.Exit(1)
```
Comment (jannfis) [inline cmd/argocd/commands/app.go:2879]: `errors.CheckError()` will already terminate the CLI, so the `os.Exit()` call is effectively dead code.

File: cmd/argocd/commands/app.go [hunk 3]
```diff
@@ -2780,3 +2797,127 @@ func NewApplicationPatchCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+
+			app, err := appIf.Get(ctx, &application.ApplicationQuery{
+				Name:         &appName,
+				Refresh:      getRefreshType(false, false),
+				AppNamespace: &appNs,
+			})
+			errors.CheckError(err)
+
+			if !app.Spec.HasMultipleSources() {
+				errors.CheckError(fmt.Errorf("app does not have multiple sources configured"))
```
Comment (jannfis) [inline cmd/argocd/commands/app.go:2896]:
```suggestion
				errors.CheckError(fmt.Errorf("Application does not have multiple sources configured"))
```

File: cmd/argocd/commands/app.go [hunk 4]
```diff
@@ -2780,3 +2797,127 @@ func NewApplicationPatchCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+			}
+
+			if len(app.Spec.GetSources()) == 1 {
+				fmt.Printf("cannot remove the only source remaining in the app")
+				os.Exit(1)
```
Comment (jannfis) [inline cmd/argocd/commands/app.go:2901]: There are two patterns for aborting CLI on an error being used throughout your change, one is:

```
fmt.Printf(...)
os.Exit(1)
```

and the other

```
errors.CheckError(fmt.Errorf(...))
```

IMO, for sake of consistency, we should only use one.

File: cmd/argocd/commands/app.go [hunk 5]
```diff
@@ -2780,3 +2797,127 @@ func NewApplicationPatchCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+			app.Spec.Sources = append(app.Spec.Sources[:source_index], app.Spec.Sources[source_index+1:]...)
+
+			_, err = appIf.UpdateSpec(ctx, &application.ApplicationUpdateSpecRequest{
+				Name:         &app.Name,
+				Spec:         &app.Spec,
+				AppNamespace: &appNs,
+			})
+			errors.CheckError(err)
+
+			fmt.Printf("application '%s' updated successfully\n", app.ObjectMeta.Name)
```
Comment (jannfis) [inline cmd/argocd/commands/app.go:2918]:
```suggestion
			fmt.Printf("Application '%s' updated successfully\n", app.ObjectMeta.Name)
```

File: cmd/argocd/commands/app.go [hunk 6]
```diff
@@ -2780,3 +2797,127 @@ func NewApplicationPatchCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+					Name:         &app.Name,
+					Spec:         &app.Spec,
+					Validate:     &appOpts.Validate,
+					AppNamespace: &appNs,
+				})
+				errors.CheckError(err)
+
+				fmt.Printf("application '%s' updated successfully\n", app.ObjectMeta.Name)
+			} else {
+				errors.CheckError(fmt.Errorf("cannot add source: application does not have spec.sources defined"))
```
Comment (jannfis) [inline cmd/argocd/commands/app.go:2851]:
```suggestion
				errors.CheckError(fmt.Errorf("Cannot add source: application does not have spec.sources defined"))
```

File: cmd/argocd/commands/app.go [hunk 7]
```diff
@@ -2780,3 +2797,127 @@ func NewApplicationPatchCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
+		Run: func(c *cobra.Command, args []string) {
+			ctx := c.Context()
+
+			if len(args) != 1 {
+				c.HelpFunc()(c, args)
+				os.Exit(1)
+			}
+
+			if source_index < 0 {
+				errors.CheckError(fmt.Errorf("index value of source cannot be less than 0"))
```
Comment (jannfis) [inline cmd/argocd/commands/app.go:2878]:
```suggestion
				errors.CheckError(fmt.Errorf("Index value of source cannot be less than 0"))
```

**Review by jannfis — APPROVED** [review]
Nice! I have a couple of nits, please check. Feel free to change before merge, or merge unchanged. However, I think consistency in UX is important enough to fix the nits.

Otherwise, LGTM.

**Comment by ishitasequeira** [comment]
Thanks @jannfis for the review. I have addressed your suggestions. Enabling auto-merge.

**Pull request merged**